KAFKA-17169: Add an EndpointsTest suite that test all of the public methods

masonyc opened the pull request:
Add tests to test all the public methods for class raft/Endpoints.java

masonyc committed:
Add tests for Raft Endpoints class

OmniaGM commented on raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java:
```diff
@@ -0,0 +1,489 @@
+final class EndpointsTest {
+    @Test
+    void testAddressWithValidEndpoint() {
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
```
BTW we have `Utils.mkMap` that can be used to define this in one statement instead of two 
```
Utils.mkMap` that can be used to define like 
`Utils.mkMap(Utils.mkEntry(ListenerName.normalised("listener"),InetSocketAddress.createUnresolved("localhost", 9092)));
```

or use simple java 

```
new HashMap<>() {{
put(ListenerName.normalised("listener"), InetSocketAddress.createUnresolved("localhost", 9092));
```

OmniaGM replied:
Also `ListenerName.normalised("listener")` and  `InetSocketAddress.createUnresolved("localhost", 9092)` seem to be duplicated in every test maybe we can define them on top of the test.

masonyc replied:
updated

OmniaGM commented on raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java:
```diff
@@ -0,0 +1,489 @@
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        Optional<InetSocketAddress> address = endpoints.address(ListenerName.normalised("listener"));
+
```
I can see that we keep calling `ListenerName.normalised("listener")` once for `endpointMap` and second time to asset the `ListenerName::value` or to pass it to `Endpoints::address`. Why not extract it to a variable at the beginning of the test?  same can go to `InetSocketAddress.createUnresolved("localhost", 9092)`

masonyc replied:
updated

OmniaGM commented on raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java:
```diff
@@ -0,0 +1,489 @@
+    void testSizeWithEmptyEndpoint() {
+        Endpoints endpoints = Endpoints.empty();
+
+        int size = endpoints.size();
```
Maybe we can combine these two lines to `assertEquals(0, Endpoints.empty().size());` and drop size variable

OmniaGM replied:
I feel `testSizeWithOneEndpoint` and `testSizeWithEmptyEndpoint` can be merged into one test and `assertEquals(1, Endpoints.fromInetSocketAddresses(endpointMap).size());`

masonyc replied:
merged to `testSize()`, also merged `testIsEmpty()`

OmniaGM commented on raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java:
```diff
@@ -0,0 +1,489 @@
+        );
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        assertFalse(endpoints.isEmpty());
```
the assert and endpoints variable line can be merged into one line `assertFalse(Endpoints.fromInetSocketAddresses(endpointMap).isEmpty());`

masonyc replied:
updated

OmniaGM commented on raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java:
```diff
@@ -0,0 +1,489 @@
+
+        Endpoints sameEndpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        assertEquals(endpoints, sameEndpoints);
```
Am not sure what the value of this test? can you explain it a bit as I recall we rarely test `hashCode` method.

masonyc replied:
Don't we need to make sure the overridden `equals()` method performs correctly? Happy to remove it is not needed.

cmccabe replied:
Seems reasonable to test `equals`

OmniaGM commented on raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java:
```diff
@@ -0,0 +1,489 @@
+    }
+
+    @Test
+    void testToString() {
```
What the value of testing `toString`? what public interfaces will `toString` impact if it change to any other string?

masonyc replied:
Was thinking it is a good idea to make sure it conforms to the overridden format. Happy to remove it if needed.

OmniaGM replied:
I don't see `toString` has any impact to be tested in the context of this class. We do test `toString` for example for metrics related code because it has an impact but this one not that much. If you checked the code you will see that we don't actively testing `toString`

masonyc replied:
removed

OmniaGM commented on raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java:
```diff
@@ -0,0 +1,489 @@
+        assertEquals(1, leaderEndpoints.size());
+
+        Iterator<BeginQuorumEpochRequestData.LeaderEndpoint> iterator = leaderEndpoints.iterator();
+        BeginQuorumEpochRequestData.LeaderEndpoint endpoint = iterator.next();
```
maybe better name here is `leaderEndpoint`? as `endpoints` above was referring to `Endpoints`

masonyc replied:
updated

OmniaGM commented on raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java:
```diff
@@ -0,0 +1,489 @@
+
+        BeginQuorumEpochRequestData.LeaderEndpointCollection leaderEndpoints = endpoints.toBeginQuorumEpochRequest();
+
+        assertEquals(0, leaderEndpoints.size());
```
`leaderEndpoints` in the assert can be replaced by its value `endpoints.toBeginQuorumEpochRequest()`

masonyc replied:
updated

OmniaGM commented on raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java:
```diff
@@ -0,0 +1,489 @@
+    void testEmpty() {
+        Endpoints emptyEndpoints = Endpoints.empty();
+
+        assertEquals(Endpoints.fromInetSocketAddresses(Collections.emptyMap()), emptyEndpoints);
```
What is the value of this test if we are testing `Endpoints::size`?

masonyc replied:
My thought was we want to test around if object returned from `empty()` matches the object from `initialization with Collection.emptyMap()` . Happy to remove it if you think its redundant.

OmniaGM replied:
I feel it is redundant

masonyc replied:
Removed

OmniaGM commented on raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java:
```diff
@@ -0,0 +1,489 @@
+
+    @Test
+    void testFromVotersRecordEndpointsWithEmptyEndpoint() {
+        List<VotersRecord.Endpoint> votersEndpoints = new ArrayList<>();
```
Can't `votersEndpoints` be `Collections.emptyList` instead? Then  `createdEndpoints` can be `Endpoints.fromVotersRecordEndpoints(Collections.emptyList());`

masonyc replied:
updated

OmniaGM reviewed (commented):
Thanks for the PR, I left few nits

masonyc committed:
update EndpointsTest to use Utils.mkMap and other pr comments

Changed map initialization to Utils.mkMap & Utils.mkEntry
Extracted common listener and socker address to variable
Simplified some tests

masonyc commented:
Thanks for the feedback @OmniaGM , I have updated code to address most issues in comments. Regarding hashcode() & toString() methods, I feel like there is a need to make sure they work as we expected. Happy to remove them if I am on the wrong direction.

masonyc committed:
fix import order

masonyc committed:
remove tests on toString & empty

masonyc committed:
Merge branch 'apache:trunk' into KAFKA-17169-Raft-Endpoints-tests

masonyc commented:
Hi @jsancio @OmniaGM , could you please have a look at the updated pr when free? Thanks.

OmniaGM commented on raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java:
```diff
@@ -0,0 +1,367 @@
+    void testVotersRecordEndpointsWithEmptyEndpoint() {
+        Endpoints endpoints = Endpoints.empty();
+
+        assertFalse(endpoints.votersRecordEndpoints().hasNext());
```
This test can be one line `assertFalse(Endpoints.empty().votersRecordEndpoints().hasNext());`

OmniaGM commented on raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java:
```diff
@@ -0,0 +1,367 @@
+    void testToBeginQuorumEpochRequestWithEmptyEndpoint() {
+        Endpoints endpoints = Endpoints.empty();
+
+        assertEquals(0, endpoints.toBeginQuorumEpochRequest().size());
```
same `assertEquals(0, Endpoints.empty().toBeginQuorumEpochRequest().size());`

OmniaGM reviewed (commented):
I had a second look and left couple of nits, but overall it looks better now. Thanks

masonyc committed:
Compact tests to one liner

masonyc commented:
Thanks for the feedback @OmniaGM , updated to address the latest comments as well.

masonyc commented:
Hi @OmniaGM @jsancio  please take another look when free, thanks.

OmniaGM reviewed (approved):
thanks for the updates, LGTM!

jsancio commented:
Thanks for the review @OmniaGM and the PR @masonyc . I take a look tomorrow.

masonyc commented:
Thanks @jsancio @OmniaGM for looking into it. Please let me know if I need to update anything.

cmccabe commented:
Thanks, @masonyc . This looks good! And thanks to @OmniaGM for reviewing the first versions. LGTM.
